### ISSUE TITLE
IMAP SMTP connection problem with Exchange 2010

### ISSUE THREAD

**Issue by InfopactMLoos** [Foundry376/Mailspring#130]
I tried to enable the IMAP provider but I ran into issues somehow. I tried to connect to my works exchange server through IMAP protocol, which should work as it does so with other clients.

It seems the Incoming mail (IMAP) is ok, however the SMTP part apparently fails. I looked a bit into SMTP protocol and as far as I understand it seems to go as planned but then abruptly aborts.  SMTP code 250 should be expected and ok. The only error I see is the message: `mailsmtp Last Error Code: 17`. I cannot seem to find any information regarding this error code so any help is appreciated!

##### Are there any related issues?

I first ran into issue #123. This however seems to be another issue so I created this new ticket.

##### What operating system are you using?
Ubuntu 16.04

##### What version of Mailspring are you using?
1.0.3-f82c8836 **EDIT: still broken on 1.0.7-664204d8**

--

**Bug?**
##### Do you have any third-party plugins installed? If so, which ones?
No

##### Is the issue related to a specific email provider (Gmail, Exchange, etc.)?
IMAP (with Exchange server 2010)

##### Is the issue reproducible with a particular attachment, message, signature, etc?

Reproduce steps are hard to supply as it seems something related to my setup. 
This is my error log, I have redacted some info.
```
----------IMAP----------
* OK The Microsoft Exchange IMAP4 service is ready.
1 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=NTLM AUTH=GSSAPI AUTH=PLAIN UIDPLUS CHILDREN IDLE NAMESPACE LITERAL+
1 OK CAPABILITY completed.
2 LOGIN "MYUSERNAME" *********
2 OK LOGIN completed.
3 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=NTLM AUTH=GSSAPI AUTH=PLAIN UIDPLUS CHILDREN IDLE NAMESPACE LITERAL+
3 OK CAPABILITY completed.
4 NAMESPACE
* NAMESPACE (("" "/")) NIL NIL
4 OK NAMESPACE completed.
5 LIST "" "*"
* LIST (\HasNoChildren) "/" Agenda
* LIST (\HasNoChildren) "/" Concepten
* LIST (\HasChildren) "/" Contactpersonen
* LIST (\Marked \HasNoChildren) "/" "Contactpersonen/Lync Contacts"
* LIST (\HasNoChildren) "/" "Conversation History"
* LIST (\HasNoChildren) "/" Drafts
* LIST (\HasNoChildren) "/" "Infected Items"
* LIST (\HasNoChildren) "/" Journaal
* LIST (\HasNoChildren) "/" Junk
* LIST (\HasNoChildren) "/" Notities
* LIST (\HasNoChildren) "/" "Ongewenste e-mail"
* LIST (\Marked \HasChildren) "/" INBOX
5 OK LIST completed.


----------SMTP----------
220 SERVERDOMAIN Microsoft ESMTP MAIL Service ready at Wed, 11 Oct 2017 15:15:07 +0200
EHLO MYPC
250-SERVERDOMAIN Hello [SERVERIP]
250-SIZE 104857600
250-PIPELINING
250-DSN
250-ENHANCEDSTATUSCODES
250-STARTTLS
250-AUTH GSSAPI NTLM
250-8BITMIME
250-BINARYMIME
250 CHUNKING
STARTTLS
220 2.0.0 SMTP server ready
EHLO MYPC
250-SERVERDOMAIN Hello [SERVERIP]
250-SIZE 104857600
250-PIPELINING
250-DSN
250-ENHANCEDSTATUSCODES
250-AUTH GSSAPI NTLM LOGIN
250-8BITMIME
250-BINARYMIME
250 CHUNKING

SMTP Last Response Code: 250

SMTP Last Response: SERVERDOMAIN Hello [SERVERIP]
SIZE 104857600
PIPELINING
DSN
ENHANCEDSTATUSCODES
AUTH GSSAPI NTLM LOGIN
8BITMIME
BINARYMIME
CHUNKING


mailsmtp Last Error Code: 17

mailsmtp Last Error Location: 10
```

--

**Comment by bengotow** [comment]
Hey there—this looks like a certificate error. The logs aren't much use because the client hangs up due to the certificate error before even trying to log in. Can you look at https://github.com/Foundry376/Mailspring/issues/144 and see if the file paths that are referenced there are valid for you? I'm surprised this is happening on Ubuntu actually.

**Comment by thob** [comment]
hmm, in `etc/ssl/certs` are also links to `/usr/share/ca-certificates/mozilla/`

**Comment by bengotow** [comment]
That's probably unrelated. Can you see if any of these paths are valid on your system? Mailspring needs to be able to find the `crt` or `pem` file in one of the paths below. If it's somewhere else, we'll need to update it.

```
        // Debian, Ubuntu, Arch: maintained by update-ca-certificates
        "/etc/ssl/certs/ca-certificates.crt",
        // Red Hat 5+, Fedora, Centos
        "/etc/pki/tls/certs/ca-bundle.crt",
        // Red Hat 4
        "/usr/share/ssl/certs/ca-bundle.crt",
        // FreeBSD (security/ca-root-nss package)
        "/usr/local/share/certs/ca-root-nss.crt",
        // FreeBSD (deprecated security/ca-root package, removed 2008)
        "/usr/local/share/certs/ca-root.crt",
        // FreeBSD (optional symlink)
        // OpenBSD
        "/etc/ssl/cert.pem",
```

**Comment by thob** [comment]
ca-certificates.crt is in `/etc/ssl/certs`, no cert.pem

**Comment by InfopactMLoos** [comment]
Sorry been to busy at work today, hopefully I can check tomorrow.

**Comment by InfopactMLoos** [comment]
Certs seems to be at the given ubuntu path as expected: 

> ls -l /etc/ssl/certs/ca-certificates.crt
-rw-r--r-- 1 root root 233394 okt  3 19:23 /etc/ssl/certs/ca-certificates.crt

**Comment by LiHRaM** [comment]
Been having the same issue with school and work accounts. Running Ubuntu, and had no trouble finding the cert file at the given paths.

**Comment by thob** [comment]
let me add that I had no issues with Linux Mint 18.2 (Mate) and Ubuntu 16.04, only on LM 17.3 i get the error

**Comment by InfopactMLoos** [comment]
Then I am suspecting that it does not have to do with Ubuntu versions. Did you get the same mailsping versions to work with mint 18.2  and ubuntu 16.04 as you had fail with mint 17.3?

**Comment by InfopactMLoos** [comment]
@bengotow is it possible to download older build versions?

**Comment by DaichiUeura** [comment]
Probably I faced the same issue on macOS when trying to connect to my company's Exchange server via IMAP and SMTP protocol. The app said 'Authentication Error - Check your username and password. (SMTP)View Log'. In my case, the SMTP server doesn't require authentication.

At first, would it be possible to not configure SMTP?

**Comment by thomaswieland** [comment]
I can confirm what DaichiUeura is telling. IMAP (Exchange) works fine, but SMTP-Setup against an Unix/Linux SMTP Server (no auth, white-listed source IP address) prevents me from finishing setup. Error message is the same (mailsmtp Last Error Code: 17).

**Comment by DaichiUeura** [comment]
`mailsmtp Last Error Code: 17` and `mailsmtp Last Error Location: 10` might appear in several different cases(no authentication is required in my case but it seems certificate stuff is being discussed here)? If so, I should have filed a new issue.

**Comment by Franky123C** [comment]
`mailsmtp Last Error Code: 17` `mailsmtp Last Error Location: 10`
Same error on Arch Linux. Version: 1.0.6-9448c884

**Comment by InfopactMLoos** [comment]
@DaichiUeura maybe, @bengotow pointed me towards certificates, however this has not lead me to any conclusive results as of now. Other people who added to my report have all reported different variables in their exposure to the same error message. This makes me think that the error code is a very common one. I would like to do more detailed research to better pin point the exact cause. However I have no idea where to check further, my knowledge of the codebase is practically none and my knowledge of the SMTP protocol is rudimentary. If anyone with better knowledge knows where to check further I would gladly do so, any help appreciated!

**Comment by aia832003** [comment]
I am having similar issues on latest version running on Fedora 26.

**Comment by InfopactMLoos** [comment]
Ok I followed the following tutorial to test the connection from the command line: [https://halon.io/blog/how-to-test-smtp-servers-using-the-command-line/](https://halon.io/blog/how-to-test-smtp-servers-using-the-command-line/). Everything is working fine with the exact same credentials and server info as I use in Mailspring so I think it is save to assume there is something bugged in Mailspring. @bengotow is there anything I can supply to help fix this issue?  Please let me know because I would really like to use this client, I loved Nylas so I must have the same love for its offspring :).

**Comment by KalSat** [comment]
I have the same problem on Ubuntu 17.10, but Geary is working.

**Comment by InfopactMLoos** [comment]
@bengotow, please some feedback on my questions? I would happily fix it with a pull request but no idea where to begin. Where does the IMAP authentication take place?

**Comment by aia832003** [comment]
This seems to be fixed in the latest release for me using Zoho Mail.

Andrew Allen

**Comment by InfopactMLoos** [comment]
@bengotow, "Allow Insecure SSL" makes no difference. As I said using the `openssl` or `swaks` tools to manually login works so I see no problems in certificates, also as other people have it working on Ubuntu and I don't have anything custom on my distro (using LTS since its a work laptop) I don't see any reason Mailspring should have trouble finding certificates.

@aia832003, your issue is not the same (although the error message might be the same but it is a generic one). So no **This seems to be fixed** is incorrect for this issue.

**Comment by fr3dm4n** [comment]
I can confirm this error using my own mail server. I have no problems using Thunderbird or Geary. But Mailspring wont connect to my SMTP-Server.

Cencosored log:
```
----------SMTP----------
220 mail.server.tld ESMTP Postfix (Ubuntu)
EHLO localmachine
250-mail.server.tld
250-PIPELINING
250-SIZE 26214400
250-ETRN
250-STARTTLS
250-ENHANCEDSTATUSCODES
250-8BITMIME
250 DSN
STARTTLS
er
220 2.0.0 Ready to start TLS
EHLO localmachine
250-mail.server.tld
250-PIPELINING
250-SIZE 26214400
250-ETRN
250-AUTH DIGEST-MD5 CRAM-MD5 PLAIN LOGIN
250-ENHANCEDSTATUSCODES
250-8BITMIME
250 DSN

SMTP Last Response Code: 250

SMTP Last Response: mail.server.tld
PIPELINING
SIZE 26214400
ETRN
AUTH DIGEST-MD5 CRAM-MD5 PLAIN LOGIN
ENHANCEDSTATUSCODES
8BITMIME
DSN


mailsmtp Last Error Code: 17

mailsmtp Last Error Location: 10
```

**Comment by InfopactMLoos** [comment]
Since my company migrated to office 365 I don't have any problem connecting mailspring to my mailbox anymore. This makes debugging this issue a bit harder and since support on this issue is also non-existing I will just close it.

**Comment by MarkPash** [comment]
I have the exact same issue, and if it helps, it's the same on windows as well. I can help test this if needed.

**Comment by InfopactMLoos** [comment]
Since I have no mailserver to reproduce this problem with anymore, no clue where to look for more debugging information, no way to differentiate between all the different issues above and mostly because there is no support or at the least pointers on where to look in the source code.... I closed this issue and feel no need to pursue this any further. If you want help to fix the issue please create a new issue, best of luck.

**Comment by ltjdym** [comment]
Same problem. Would have loved to keep using mailspring otherwise...

**Comment by ltjdym** [comment]
BTW, can use newton mail without any problems... Maybe mailspring only works with outlook 365 and not outlook.com ? Or outlook exchange?

**Comment by ltjdym** [comment]
----------IMAP----------
* OK The Microsoft Exchange IMAP4 service is ready. [QgBOADYAUABSADEAMABDAEEAMAAwADMAOQAuAG4AYQBtAHAAcgBkADEAMAAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]
ذ�S1 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
1 OK CAPABILITY completed.
2 LOGIN "xxxxx@xxxx.org" *********
2 OK LOGIN completed.
3 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
3 OK CAPABILITY completed.
y�xm�4 NAMESPACE
4 BAD User is authenticated but not connected.
�M�S

**Comment by ltjdym** [comment]
----------IMAP----------
* OK The Microsoft Exchange IMAP4 service is ready. [TQBXAEgAUABSADIAMgBDAEEAMAAwADEANwAuAG4AYQBtAHAAcgBkADIAMgAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]
�oIY1 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
1 OK CAPABILITY completed.
2 LOGIN "xxxxx@xxxx.org" *********
2 OK LOGIN completed.
3 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
3 OK CAPABILITY completed.
����4 NAMESPACE
4 BAD User is authenticated but not connected.

�dY

**Comment by ltjdym** [comment]
im unable to turn on IMAP in my outlook.com account. I think that's the culprit.

**Comment by InfopactMLoos** [comment]
This issue is closed and not being pursuit anymore, please open a new issue.

**Comment by fr3dm4n** [comment]
@thomaswieland I just tried to reproduce it using my own mail server. Do you use Linux? Because only in Linux (Fedora) I am able to reproduce it. In Windows everything works well.

**Comment by thomaswieland** [comment]
@fr3dm4n I am using Mailspring on Windows 7 Enterprise x64. As SMTP Server I failed to connect
- Exchange Server 2013 (Windows) with NTLM authentication
- Postfix (Unix) without authentication (Source IP whitelist)
- and initially hMailServer (Windows) with PLAIN and LOGIN

hMailServer had PLAIN and LOGIN authentication enabled at first; I couldn´t set up Mailspring, I got stuck with the error messages (see above). After disabling PLAIN auth on hMailServer (offering only LOGIN auth), I could set up Mailspring successfully. Does that answer your question?
